src/hotspot/share/runtime/handles.hpp

```diff
@@ -288,6 +288,9 @@ class NoHandleMark: public StackObj {
 };
 
 
+// ResetNoHandleMark is called in a context where there is an enclosing
+// NoHandleMark. A thread in _thread_in_native must not create handles so
+// this is used when transitioning via ThreadInVMfromNative.
```

Comment:
This comment still doesn't clearly (to me) explain the use of RNHM. Basically it is intended to "disable" a NHM further up the stack but why do we not force the caller up the stack to correctly scope the NMH? This seems to be a way to allow sloppy coding further up the stack, or worse it allows code to be called that might use Handles completely unexpectedly wrt the code up the stack. Should a NHM not mean "this code and all code called from it must not use Handles"?

And can't this be used in a range of contexts not just this ThreadInVMFromNative transition?

Thanks,
David

Reply:
Yes so the NHM does mean "this code and all code called from this must not use handles" which is why it appears in the JNI_LEAF and JVM_LEAF.  Handles and oops are not safe in VM code whose state is _thread_in_native.
So the interesting scenario here is:

JNI_LEAF(some function) {
  do some things that don't allow oops or Handles (the NHM will cause an assertion to help make sure this is the case)
  // Oh rats, we need to go into the VM maybe conditionally
  ThreadInVMFromNative tivm;
   // Now handles are cool.
}
This change embeds the ResetNoHandleMark in the ThreadInVMFromNative so that all the callers don't require it.

The caller up the stack is the JNI_LEAF generally where we generally want NoHandleMark.

The transitions from Java to VM don't require this because my n+1 change lifts the no Handle restrictions.  This is the only case where this pattern is needed in the callers.  My goal with this is to reduce the intellectual noise required to read this and figure out why it's needed that is littered around the source code.

I'm not sure what to say in this comment that explains this better, please suggest better wording.